hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/MountVolumeInfo.java

```diff
@@ -102,9 +116,28 @@ boolean addVolume(FsVolumeImpl volume) {
     return true;
   }
 
-
   void removeVolume(FsVolumeImpl target) {
     storageTypeVolumeMap.remove(target.getStorageType());
+    capacityRatioMap.remove(target.getStorageType());
+  }
+
+  /**
+   * Set customize capacity ratio for a storage type.
+   * Return false if the value is too big.
+   */
+  boolean setCapacityRatio(StorageType storageType,
+      double capacityRatio) {
+    double leftover = 1;
+    for (Map.Entry<StorageType, Double> e : capacityRatioMap.entrySet()) {
+      if (e.getKey() != storageType) {
+        leftover -= e.getValue();
+      }
+    }
+    if (leftover < capacityRatio) {
+      return false;
+    }
+    capacityRatioMap.put(storageType, capacityRatio);
```

Comment:
Do we allow overwrite an existing capacity ratio of the storage type?

Reply:
I think later it will be convenient when we add a feature to update the capacity ratio without restarting datanode. It should be pretty harmless for now.

Reply:
Is it possible that this setCapacityRatio call is triggered by refreshVolumes op? In that case if we do not reload capacity ratio configuration for refreshVolumes, we can have inconsistency here. So I think we need to make sure this new feature works well along with refreshVolumes. What do you think?

Reply:
Yes, I agree. This is a good point. We need to refresh the capacity ratio as well when calling refreshVolumes to make this a complete feature. Let me spend some time on it.